lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1439,44 +1489,69 @@ class HSAQueue final : public KalmarQueue
         unsigned int physical_count = device->get_compute_unit_count();
         assert(physical_count > 0);
 
-        std::vector<uint32_t> cu_arrays;
         uint32_t temp = 0;
         uint32_t bit_index = 0;
 
         // If cu_mask.size() is greater than physical_count, igore the rest.
         int iter = cu_mask.size() > physical_count ? physical_count : cu_mask.size();
 
-        for(auto i = 0; i < iter; i++) {
-            temp |= (uint32_t)(cu_mask[i]) << bit_index;
 
-            if(++bit_index == 32) {
-                cu_arrays.push_back(temp);
-                bit_index = 0;
-                temp = 0;
+        {
+            std::lock_guard<std::mutex> (this->qmutex);
+
+
+            this->cu_arrays.clear();
+
+            for(auto i = 0; i < iter; i++) {
+                temp |= (uint32_t)(cu_mask[i]) << bit_index;
+
+                if(++bit_index == 32) {
+                    this->cu_arrays.push_back(temp);
+                    bit_index = 0;
+                    temp = 0;
+                }
             }
-        }
 
-        if(bit_index != 0) {
-            cu_arrays.push_back(temp);
-        }
+            if(bit_index != 0) {
+                this->cu_arrays.push_back(temp);
+            }
 
-        // call hsa ext api to set cu mask
-        hsa_status_t status = hsa_amd_queue_cu_set_mask(commandQueue, cu_arrays.size(), cu_arrays.data());
-        if(HSA_STATUS_SUCCESS == status)
-            return true;
-        else
-            return false;
+
+            // Apply the new cu mask to the hw queue:
+            return (rocrQueue->setCuMask(this) == HSA_STATUS_SUCCESS);
+
+        }
     }
 
     // enqueue a barrier packet
     std::shared_ptr<KalmarAsyncOp> EnqueueMarker() override {
+
         hsa_status_t status = HSA_STATUS_SUCCESS;
 
         // create shared_ptr instance
         std::shared_ptr<HSABarrier> barrier = std::make_shared<HSABarrier>();
 
         // enqueue the barrier
-        status = barrier.get()->enqueueAsync(this);
+        status = barrier.get()->enqueueAsync(this, false/*agent-flush*/);
+        STATUS_CHECK(status, __LINE__);
```

Comment:
I think this is changing the semantics of EnqueueMarker (and also EnqueueMarkerWithDependency) by changing it from system scope to agent scope fence.  I think we expose these through the hc::create_marker API so this will break user code the marker was used as the synchronization mechanism between the host and the device?